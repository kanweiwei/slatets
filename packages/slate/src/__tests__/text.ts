import Text from "../models/text";
import Key from "../utils/key-utils";
import Leaf from "../models/leaf";
import Mark from "../models/mark";
describe("test Text", () => {
  test("constructor", () => {
    const obj1 = {};
    const t1 = new Text(obj1);
    expect(t1).toBeInstanceOf(Text);
    expect(t1.key).toBeInstanceOf(Key);
    expect(t1.leaves).toEqual([]);

    const obj2 = {
      leaves: [
        {
          text: "abc",
          marks: [{ type: "bold" }],
        },
      ],
    };
    const t2 = new Text(obj2);
    expect(t2.leaves.length).toBe(1);
    expect(t2.leaves[0]).toBeInstanceOf(Leaf);
    expect(t2.leaves[0].text).toBe("abc");
  });

  test("create", () => {
    const t1 = Text.create("abc");
    expect(t1.leaves.length).toBe(1);
    expect(t1.leaves[0]).toBeInstanceOf(Leaf);
    expect(Text.create(t1) === t1).toBeTruthy();

    const obj = {
      leaves: [
        {
          text: "abc",
          marks: [{ type: "bold" }],
        },
      ],
    };
    const t2 = Text.create(obj);
    expect(t2).toBeInstanceOf(Text);
    expect(Text.create()).toBeInstanceOf(Text);
    expect(t2).toMatchSnapshot();
  });

  test("createList", () => {
    const elements = [
      {
        leaves: [
          {
            text: "hello",
          },
        ],
      },
      {
        leaves: [
          {
            text: "world",
          },
        ],
      },
    ];
    const texts = Text.createList(elements);
    expect(texts.length).toBe(2);
    expect(texts.every((n) => Text.isText(n))).toBeTruthy();
    expect(texts).toMatchSnapshot();
  });

  test("fromJSON", () => {
    const leaves = [
      {
        text: "abc",
        marks: [
          {
            type: "bold",
          },
        ],
      },
    ];
    const t = Text.fromJSON({ leaves });
    expect(t).toBeInstanceOf(Text);
    expect(t.getText()).toBe("abc");
    expect(t.key).toBeInstanceOf(Key);
    expect(t).toMatchSnapshot();
  });

  test("isText", () => {
    const t = Text.create("abc");
    expect(t).toBeInstanceOf(Text);
    expect(Text.isText(t)).toBeTruthy();
    expect(Text.isText({})).toBeFalsy();
    expect(t).toMatchSnapshot();
  });

  test("isTextList", () => {
    const ts = ["A", "B"].map(Text.create);
    expect(Text.isTextList(ts)).toBeTruthy();
    expect(Text.isTextList([])).toBeTruthy();
    expect(Text.isTextList([{}])).toBeFalsy();
  });

  test("getText", () => {
    const t = Text.create("hello world");
    expect(t.getText()).toBe("hello world");
  });

  test("isEmpty", () => {
    const t = Text.create("");
    expect(t.isEmpty).toBeTruthy();
  });

  test("search leaf at offset", () => {
    const t = Text.create({
      leaves: [
        {
          text: "hello",
        },
        {
          text: "world",
        },
      ],
    });
    let leaf = t.searchLeafAtOffset(5).leaf;
    expect(leaf?.text).toBe("hello");
    expect(t.searchLeafAtOffset(5)).toMatchSnapshot();
    leaf = t.searchLeafAtOffset(6).leaf;
    expect(leaf?.text).toBe("world");
    expect(t.searchLeafAtOffset(6)).toMatchSnapshot();
  });

  test("addMark", () => {
    let t = Text.create({
      leaves: [
        {
          text: "hello",
        },
        {
          text: "world",
        },
      ],
    });
    t = t.addMark(3, 2, Mark.create("bold"));
    expect(t).toMatchObject({
      object: "text",
      leaves: [
        {
          text: "hel",
          marks: [],
          object: "leaf",
        },
        {
          text: "lo",
          object: "leaf",
          marks: [
            {
              type: "bold",
              object: "mark",
              data: {},
            },
          ],
        },
        {
          text: "world",
          object: "leaf",
          marks: [],
        },
      ],
    });
    expect(t).toMatchSnapshot();
  });

  test("addMarks", () => {
    let t = Text.create({
      leaves: [
        {
          text: "hello",
        },
        {
          text: "world",
        },
      ],
    });
    t = t.addMarks(3, 2, [Mark.create("bold"), Mark.create("underline")]);
    expect(t).toMatchObject({
      object: "text",
      leaves: [
        {
          text: "hel",
          marks: [],
          object: "leaf",
        },
        {
          text: "lo",
          object: "leaf",
          marks: [
            {
              type: "bold",
              object: "mark",
              data: {},
            },
            {
              type: "underline",
              object: "mark",
              data: {},
            },
          ],
        },
        {
          text: "world",
          object: "leaf",
          marks: [],
        },
      ],
    });
    expect(t).toMatchSnapshot();
  });

  test("getLeaves without decorations", () => {
    const t = Text.create({
      leaves: [
        {
          text: "hello",
        },
        {
          text: " world",
        },
      ],
    });
    const leaves = t.getLeaves();
    expect(leaves).toMatchObject([
      {
        text: "hello",
      },
      {
        text: " world",
      },
    ]);
  });

  test("getLeaves with decorations", () => {});
});
