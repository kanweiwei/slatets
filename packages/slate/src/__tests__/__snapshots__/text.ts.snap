// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Text addMark 1`] = `
Object {
  "leaves": Array [
    Object {
      "marks": Array [],
      "object": "leaf",
      "text": "hel",
    },
    Object {
      "marks": Array [
        Object {
          "data": Object {},
          "object": "mark",
          "type": "bold",
        },
      ],
      "object": "leaf",
      "text": "lo",
    },
    Object {
      "marks": Array [],
      "object": "leaf",
      "text": "world",
    },
  ],
  "object": "text",
}
`;

exports[`test Text addMarks 1`] = `
Object {
  "leaves": Array [
    Object {
      "marks": Array [],
      "object": "leaf",
      "text": "hel",
    },
    Object {
      "marks": Array [
        Object {
          "data": Object {},
          "object": "mark",
          "type": "bold",
        },
        Object {
          "data": Object {},
          "object": "mark",
          "type": "underline",
        },
      ],
      "object": "leaf",
      "text": "lo",
    },
    Object {
      "marks": Array [],
      "object": "leaf",
      "text": "world",
    },
  ],
  "object": "text",
}
`;

exports[`test Text create 1`] = `
Object {
  "leaves": Array [
    Object {
      "marks": Array [
        Object {
          "data": Object {},
          "object": "mark",
          "type": "bold",
        },
      ],
      "object": "leaf",
      "text": "abc",
    },
  ],
  "object": "text",
}
`;

exports[`test Text createList 1`] = `
Array [
  Object {
    "leaves": Array [
      Object {
        "marks": Array [],
        "object": "leaf",
        "text": "hello",
      },
    ],
    "object": "text",
  },
  Object {
    "leaves": Array [
      Object {
        "marks": Array [],
        "object": "leaf",
        "text": "world",
      },
    ],
    "object": "text",
  },
]
`;

exports[`test Text fromJSON 1`] = `
Object {
  "leaves": Array [
    Object {
      "marks": Array [
        Object {
          "data": Object {},
          "object": "mark",
          "type": "bold",
        },
      ],
      "object": "leaf",
      "text": "abc",
    },
  ],
  "object": "text",
}
`;

exports[`test Text isText 1`] = `
Object {
  "leaves": Array [
    Object {
      "marks": Array [],
      "object": "leaf",
      "text": "abc",
    },
  ],
  "object": "text",
}
`;

exports[`test Text search leaf at offset 1`] = `
Object {
  "endOffset": 5,
  "index": 0,
  "leaf": Object {
    "marks": Array [],
    "object": "leaf",
    "text": "hello",
  },
  "startOffset": 0,
}
`;

exports[`test Text search leaf at offset 2`] = `
Object {
  "endOffset": 10,
  "index": 1,
  "leaf": Object {
    "marks": Array [],
    "object": "leaf",
    "text": "world",
  },
  "startOffset": 5,
}
`;
